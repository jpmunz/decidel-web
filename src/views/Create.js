import React from "react";
import { Alert, Form } from "react-bootstrap";
import ListInput from "./ListInput";
import { getAsyncRequest, createDecidelRequest } from "src/state";
import LoadingButton from "./LoadingButton";
import { Redirect } from "react-router-dom";
import { connect } from "react-redux";
import autoBind from "react-autobind";
import "./Create.scss";

export class Create extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
    autoBind(this);
  }

  handleNameChange(event) {
    this.setState({ name: event.target.value });
  }

  handleTitleChange(event) {
    this.setState({ title: event.target.value });
  }

  handleOptionsChange(options) {
    this.setState({ options: options });
  }

  handleDecidersChange(deciders) {
    this.setState({ deciders: deciders });
  }

  handleCreate(event) {
    event.preventDefault();

    const deciders = [{ name: this.state.name }];
    this.state.deciders.forEach((name, i) => {
      deciders.push({ name: name, isNext: i === 0 });
    });

    this.props.createDecidelRequest({
      requestId: this.props.requestId,
      decidel: {
        deciders: deciders,
        title: this.state.title,
        options: this.state.options.map(option => ({ title: option })),
        history: []
      }
    });
  }

  isValid() {
    return (
      this.state.name &&
      this.state.name.length &&
      this.state.title &&
      this.state.title.length &&
      this.state.deciders &&
      this.state.deciders.length &&
      this.state.deciders.every(item => item.length) &&
      this.state.options &&
      this.state.options.length > 1 &&
      this.state.options.every(item => item.length)
    );
  }

  render() {
    if (this.props.request.success) {
      return <Redirect to={"/view/" + this.props.request.response.id} />;
    }

    let alert;
    if (this.props.request.error) {
      alert = (
        <Alert variant="warning">
          There was a problem creating your Decidel. Please try again later.
        </Alert>
      );
    }

    return (
      <Form className="Create">
        <Form.Group>
          <Form.Label htmlFor="name-input">What's your name?</Form.Label>
          <Form.Control
            id="name-input"
            disabled={this.props.request.inProgress}
            onChange={this.handleNameChange}
            type="text"
            placeholder="The Decider"
          />
        </Form.Group>

        <Form.Group>
          <Form.Label htmlFor="title-input">
            What are you trying to decide?
          </Form.Label>
          <Form.Control
            id="title-input"
            disabled={this.props.request.inProgress}
            onChange={this.handleTitleChange}
            type="text"
            placeholder="What movie to watch"
          />
        </Form.Group>

        <Form.Group>
          <Form.Label>What options do you have?</Form.Label>
          <ListInput
            disabled={this.props.request.inProgress}
            numInitialOptions={2}
            onChange={this.handleOptionsChange}
            addMessage="Add another option"
          />
        </Form.Group>

        <Form.Group>
          <Form.Label>Who else is deciding?</Form.Label>
          <ListInput
            disabled={this.props.request.inProgress}
            onChange={this.handleDecidersChange}
            addMessage="Add another decider"
          />
        </Form.Group>

        <div className="footer">
          {alert}
          <LoadingButton
            className="create-button"
            type="submit"
            loading={this.props.request.inProgress}
            disabled={!this.isValid()}
            onClick={this.handleCreate}
            variant="primary"
            text="Create"
          />
        </div>
      </Form>
    );
  }
}

export default connect(
  (state, props) => ({
    request: getAsyncRequest(state, props.requestId)
  }),
  { createDecidelRequest }
)(Create);
