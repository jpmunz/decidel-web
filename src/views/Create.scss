@import "mixins";

.Create {
  width: 60%;
  margin: 50px auto;
  padding: 10px;
  @include rounded();
  @include box-shadow(10px, 10px, 20px, rgba(0, 0, 0, 0.2));

  .form-control {
    border-top: none;
    border-left: none;
    border-right: none;
    border-radius: 0;
    padding: 0;
    box-shadow: none;
    width: calc(100% - 40px);
  }

  .form-group {
    margin: 0 auto;
    width: 85%;
  }

  .form-label {
    margin-top: 30px;
    margin-bottom: 20px;
  }

  .footer {
    text-align: right;
  }

  .create-button {
    @include box-shadow(5px, 5px, 10px, rgba(0, 0, 0, 0.2));
  }
}
