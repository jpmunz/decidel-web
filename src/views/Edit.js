import React from "react";
import { Container, Row, Col } from "react-bootstrap";
import { connect } from "react-redux";
import {
  fetchCurrentUserRequest,
  fetchDecidelRequest,
  getFinalDecision,
  getAsyncRequest,
  hasDecidel,
  getTitle
} from "src/state";
import DecisionOptions from "./DecisionOptions";
import Deciders from "./Deciders";
import History from "./History";
import Loading from "./Loading";
import CurrentRound from "./CurrentRound";
import "./Edit.scss";

export class Edit extends React.Component {
  componentDidMount() {
    this.props.fetchDecidelRequest({ id: this.props.id });
    this.props.fetchCurrentUserRequest();
  }

  render() {
    let content;

    if (this.props.request.error) {
      content = (
        <>
          <Row className="justify-content-center">
            <Col className="decidel-not-found">
              <div>We couldn't find a Decidel with that id.</div>
              <div>Are you sure you entered the URL correctly?</div>
            </Col>
          </Row>
        </>
      );
    } else if (!this.props.decidelLoaded || this.props.request.inProgress) {
      return <Loading />;
    } else if (this.props.finalDecision) {
      content = (
        <>
          <Row>
            <Col>
              <div className={"final-decision-header"}>
                <div className={"final-decision-preamble"}>You decided on</div>
                <div className={"final-decision-title"}>
                  {this.props.finalDecision + "!"}
                </div>
              </div>
            </Col>
          </Row>
          <Row className="justify-content-center">
            <Col className={"col-8"}>
              <DecisionOptions id={this.props.id} />
            </Col>
          </Row>
          <Row className="justify-content-center">
            <Col className={"col-8"}>
              <History id={this.props.id} />
            </Col>
          </Row>
        </>
      );
    } else {
      content = (
        <>
          <Row>
            <Col>
              <h1>{this.props.title}</h1>
            </Col>
          </Row>
          <Row>
            <Col>
              <Row>
                <Col>
                  <CurrentRound id={this.props.id} />
                </Col>
              </Row>
              <Row>
                <Col>
                  <DecisionOptions id={this.props.id} />
                </Col>
              </Row>
            </Col>
            <Col className="col-sm">
              <Deciders id={this.props.id} />
            </Col>
          </Row>
          <Row>
            <Col>
              <History id={this.props.id} />
            </Col>
          </Row>
        </>
      );
    }

    return <Container className="Edit">{content}</Container>;
  }
}

export default connect(
  (state, props) => {
    return {
      finalDecision: getFinalDecision(state, props.id),
      request: getAsyncRequest(state, props.id),
      title: getTitle(state, props.id),
      decidelLoaded: hasDecidel(state, props.id)
    };
  },
  { fetchDecidelRequest, fetchCurrentUserRequest }
)(Edit);
