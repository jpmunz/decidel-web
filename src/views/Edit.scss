@import "mixins";

.Edit {
  h1 {
    padding-bottom: 0.5em;
    text-align: center;
  }

  .History {
    padding-top: 0.5em;
  }

  .spinner {
    text-align: center;
  }

  .final-decision-header {
    text-align: center;
    margin-bottom: 20px;

    .final-decision-preamble {
      font-size: 3em;
    }

    .final-decision-title {
      font-weight: bold;
      font-size: 3.5em;
    }
  }

  .decidel-not-found {
    text-align: center;
    font-size: 1.5em;
  }
}
