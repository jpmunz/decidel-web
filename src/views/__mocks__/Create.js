import React from "react";

export const MockCreate = jest.fn();

MockCreate.mockReturnValue(<div>Mocked Create</div>);

const mock = jest.fn().mockImplementation(MockCreate);

export default mock;
