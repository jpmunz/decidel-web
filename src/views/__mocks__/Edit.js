import React from "react";

export const MockEdit = jest.fn();

MockEdit.mockReturnValue(<div>Mocked Edit</div>);

const mock = jest.fn().mockImplementation(MockEdit);

export default mock;
