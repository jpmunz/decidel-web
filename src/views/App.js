import React from "react";
import { Switch, Route } from "react-router-dom";
import Header from "./Header";
import Home from "./Home";
import Create from "./Create";
import Edit from "./Edit";
import v4 from "uuid/v4";

class App extends React.Component {
  render() {
    return (
      <div className="App">
        <Header />
        <div className="container justify-content-center">
          <Switch>
            <Route exact path="/">
              <Home />
            </Route>
            <Route
              path="/create"
              render={() => {
                return <Create requestId={v4()} />;
              }}
            />
            <Route
              path="/view/:id"
              render={props => {
                return <Edit id={props.match.params.id} />;
              }}
            />
          </Switch>
        </div>
      </div>
    );
  }
}
export default App;
