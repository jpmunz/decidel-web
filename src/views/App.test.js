import React from "react";
import App from "./App";
import { MemoryRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import { MockCreate } from "./Create";
import { MockEdit } from "./Edit";

jest.mock("./Home");
jest.mock("./Create");
jest.mock("./Edit");

jest.mock("uuid/v4", () => ({
  __esModule: true,
  default: () => "v4id"
}));

const renderComponent = function(route) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <App />
    </MemoryRouter>
  );
};

test("Default route", () => {
  expect(renderComponent("/")).toMatchSnapshot();
});

test("Create route", () => {
  expect(renderComponent("/create")).toMatchSnapshot();
  expect(MockCreate.mock.calls).toEqual([[{ requestId: "v4id" }, {}]]);
});

test("Edit route", () => {
  expect(renderComponent("/view/23")).toMatchSnapshot();
  expect(MockEdit.mock.calls).toEqual([[{ id: "23" }, {}]]);
});
