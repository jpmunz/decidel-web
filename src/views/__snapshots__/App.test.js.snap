// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create route 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="Header"
    >
      <a
        class="created-by"
        href="https://portfolio.jpmunz.com"
      >
        By Jonathan Munz
      </a>
      <a
        href="/"
      >
        <img
          alt=""
          class="logo"
          src="logo.svg"
        />
      </a>
    </div>
    <div
      class="container justify-content-center"
    >
      <div>
        Mocked Create
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Default route 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="Header"
    >
      <a
        class="created-by"
        href="https://portfolio.jpmunz.com"
      >
        By Jonathan Munz
      </a>
      <a
        href="/"
      >
        <img
          alt=""
          class="logo"
          src="logo.svg"
        />
      </a>
    </div>
    <div
      class="container justify-content-center"
    >
      <div>
        Mocked Home
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Edit route 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="Header"
    >
      <a
        class="created-by"
        href="https://portfolio.jpmunz.com"
      >
        By Jonathan Munz
      </a>
      <a
        href="/"
      >
        <img
          alt=""
          class="logo"
          src="logo.svg"
        />
      </a>
    </div>
    <div
      class="container justify-content-center"
    >
      <div>
        Mocked Edit
      </div>
    </div>
  </div>
</DocumentFragment>
`;
